ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,810 @@
+
+-- | Apply a list of transactions short-circuiting if the mempool gets full.
+-- Emulates almost exactly the behavior of 'implTryAddTxs'.
+foldTxs ::
+  forall blk.
+  ( LedgerSupportsMempool blk
+  , BasicEnvelopeValidation blk
+  )
+  => LedgerConfig blk
+  -> TicketNo
+  -> MempoolCapacityBytes
+  -> TickedLedgerState blk ValuesMK
+  -> [GenTx blk]
+  -> ( ([MempoolAddTxResult blk], [GenTx blk])
+     , TicketNo
+     , MempoolCapacityBytes
+     , TickedLedgerState blk ValuesMK
+     )
+foldTxs cfg nextTk remainingCap initialState  =
+    go ( []
+       , nextTk
+       , getMempoolCapacityBytes remainingCap
+       , initialState
+       )
+  where
+    go (acc, tk, cap, st) [] = ((reverse acc, [])
+                               , tk
+                               , MempoolCapacityBytes cap
+                               , st
+                               )
+    go (acc, tk, cap, st) txs@(tx:next) =
+      if cap < txInBlockSize tx
+      then ((reverse acc, txs), tk, MempoolCapacityBytes cap, st)
+      else
+        let slot = case getTipSlot st of
+              Origin -> minimumPossibleSlotNo (Proxy @blk)
+              At v   -> v + 1
+        in
+        case runExcept $ applyTx cfg DoNotIntervene slot tx st of
+          Left e           ->
+            go ( MempoolTxRejected tx e:acc
+               , tk
+               , cap
+               , st
+               )
+               next
+          Right (st', vtx) ->
+            go ( MempoolTxAdded vtx:acc
+               , succ tk
+               , cap - txInBlockSize tx
+               , forgetLedgerTablesDiffsTicked st'
+               )
+               next
+
+tick ::
+  ( TickedTableStuff (LedgerState blk)
+  , ValidateEnvelope blk
+  , LedgerSupportsMempool blk
+  )
+  => LedgerConfig blk
+  -> LedgerState blk ValuesMK
+  -> TickedLedgerState blk ValuesMK
+tick cfg st =
+  zipOverLedgerTablesTicked
+    (flip rawApplyDiffs)
+    ticked
+    (projectLedgerTables st)
+  where
+    ticked = snd
+           . tickLedgerState cfg
+           . ForgeInUnknownSlot
+           . forgetLedgerTables
+           $ st
+
+{-------------------------------------------------------------------------------
+  SUT side
+-------------------------------------------------------------------------------}
+
+-- | The System Under Test
+data SUT m blk =
+  SUT
+  !(Mempool m blk)
+    -- ^ A Mempool
+  !(StrictTVar m (Tips blk))
+    -- ^ A TVar with tips. Emulates a ledger db to the extent needed by the
+    -- ledger interface.
+  deriving Generic
+
+deriving instance ( NoThunks (Mempool m blk)
+                  , NoThunks (StrictTVar m (Tips blk))
+                  ) =>  NoThunks (SUT m blk)
+
+-- | The tip of the ledger db and the one from the mempool. The ledger interface
+-- will serve the values of the latter if the mempool requests it and is
+-- reachable.
+data Tips blk = Tips {
+    ldbTip     :: !(LedgerState blk ValuesMK)
+  , mempoolTip :: !(LedgerState blk ValuesMK, Bool)
```

Comment:
Why do we need the mempool tip? I'm a bit worried that we have two mempool tips in the `SUT`.